instrumentation/ratpack-1.4/library/src/main/java/io/opentelemetry/instrumentation/ratpack/OpenTelemetryExecInterceptor.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.ratpack;
+
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import ratpack.exec.ExecInterceptor;
+import ratpack.exec.Execution;
+import ratpack.func.Block;
+
+final class OpenTelemetryExecInterceptor implements ExecInterceptor {
+
+  static final ExecInterceptor INSTANCE = new OpenTelemetryExecInterceptor();
+
+  @Override
+  public void intercept(Execution execution, ExecType type, Block continuation) throws Exception {
+    Context otelCtx = execution.maybeGet(Context.class).orElse(null);
+    if (otelCtx != null) {
+      // Execution already has a context, this is an asynchronous resumption and we need to make
+      // the context current.
+      try (Scope ignored = otelCtx.makeCurrent()) {
+        continuation.execute();
+      }
+      return;
```

Comment:
Won't returning here cause the scope to never be closed? Doesn't `OpenTelemetryServerHandler` always add both `Context` and `Scope`? If they're both present then the second part of this method won't execute.

Reply:
The handler is always invoked as part of the execution of a continuation. So initially, there is no context or scope. After continuation.execute here there will be both context and scope. And then we close the scope, leaving the context. If the execution is reused then the context is remounted above.

I was hoping my comment below captured this is there a way I can make it clearer?

Reply:
Maybe it's enough if I flip the order of the ifs so the order matches execution order?

Reply:
So when there's both context and scope in the execution then this block `if (otelCtx != null) { /* ... */ return; }` will execute and return from the `ExecInterceptor` before closing & removing scope. Maybe I'm missing or misunderstanding something, but wouldn't this class close the scope only in a "no context & scope" situation? While there seem to be only 3 options possible ("no context & no scope" - before handler, "context & scope" - after handler but before execution finishes, "context & no scope" - after execution)

Reply:
> Maybe it's enough if I flip the order of the ifs so the order matches execution order?

I think it'd be enough - that's what confused me here

Reply:
Ok - don't think it's possible to get any more descriptive than the methods I extracted too :)

Reply:
Ahh I finally understood that - I didn't realize that `ExecInterceptor` basically wraps over the `Handler`, now it makes sense. Thanks!